### PR TITLE
Feature/ch53/implement semantic analysis

### PR DESCRIPTION
**Pull request by adityagoel4512**
### Clubhouse Ticket
[ch53](https://app.clubhouse.io/manim-dsl/story/53/implement-semantic-analysis)

### Description
+ Semantic Analysis
+ Type inference
+ Error handling

### Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### How Has This Been Tested?

We have been testing using the example file and debugger. Tests are a WIP.

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/main/kotlin/com/manimdsl/ManimDSLParser.kt [hunk 1]
```diff
@@ -2,13 +2,18 @@ package com.manimdsl
...
 import com.manimdsl.errorhandling.ErrorHandler
 import com.manimdsl.errorhandling.syntaxerror.SyntaxErrorListener
 import com.manimdsl.errorhandling.syntaxerror.SyntaxErrorStrategy
+import com.manimdsl.frontend.SymbolTable
 import org.antlr.v4.runtime.CharStreams
 import org.antlr.v4.runtime.CommonTokenStream
+import org.antlr.v4.runtime.atn.PredictionMode
 import java.io.InputStream
 
+val SYNTAX_ERROR = 1;
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/ManimDSLParser.kt:15]: Rebase on master and this should disappear. I removed it before

File: src/main/kotlin/com/manimdsl/ManimDSLParser.kt [hunk 2]
```diff
@@ -29,14 +34,19 @@ class ManimDSLParser(private val input: InputStream) {
...
-
+        // Speeds up parser with no backtracking as our grammar is quite simple
+        parser.interpreter.predictionMode = PredictionMode.SLL
         parser.errorHandler = SyntaxErrorStrategy()
         parser.removeErrorListeners()
         parser.addErrorListener(SyntaxErrorListener())
-
-        val program = parser.program();
-
+//        parser.addErrorListener(SyntaxErrorListener())
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/ManimDSLParser.kt:42]: Same here

File: src/main/kotlin/com/manimdsl/ManimParserVistor.kt [hunk 3]
```diff
@@ -2,61 +2,140 @@ package com.manimdsl
...
-        return super.visitMethodCall(ctx)
+    override fun visitMethodCall(ctx: ManimParser.MethodCallContext): MethodCallNode {
+        // Type signature of methods to be determined by symbol table
+        val arguments: List<ExpressionNode> = visitArgumentList(ctx.arg_list() as ManimParser.ArgumentListContext?).arguments
+        val identifier = ctx.IDENT(0).symbol.text
+        val methodName = ctx.IDENT(1).symbol.text
+
+        semanticAnalyser.undeclaredIdentifier(currentSymbolTable, identifier, ctx)
+        semanticAnalyser.failIfNotDataStructure(currentSymbolTable, identifier, ctx)
+        semanticAnalyser.notValidMethodNameForDataStructure(currentSymbolTable, identifier, methodName, ctx)
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/ManimParserVistor.kt:73]: Rename above to sound more conditional and not absolute e.g. undeclaredIdentifier -> undeclaredIdentifierCheck

File: src/main/kotlin/com/manimdsl/errorhandling/semanticerror/SemanticErrorHandling.kt [hunk 4]
```diff
@@ -1,3 +1,45 @@
...
+}
+
+fun typeOfArgsInMethodCallError(dataStructureType: String, method: String, argType: String, argName:String, ctx: ParserRuleContext){
+    // To modify once nex version is merged
+    addSemanticError("$method method on $dataStructureType does not accept argument $argName of type $argType", getErrorLinePos(ctx))
+}
+
+/* Helper function that returns line and character position for errors */
+private fun getErrorLinePos(ctx: ParserRuleContext): String {
+    val line = ctx.getStart().line
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/errorhandling/semanticerror/SemanticErrorHandling.kt:42]: Is this not duplicated in syntax error handler

File: src/main/kotlin/com/manimdsl/errorhandling/syntaxerror/SyntaxErrorHandling.kt [hunk 5]
```diff
@@ -11,5 +11,6 @@ fun extraneousInputError(token: String, line: Int, char: Int, underlinedError: S
 }
 
 fun otherError(msg: String, token: String, line: Int, char: Int, underlinedError: String) {
+    println(token)
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/errorhandling/syntaxerror/SyntaxErrorHandling.kt:14]: Debugging print left behind

File: src/main/kotlin/com/manimdsl/frontend/AbstractSyntaxTree.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+object NumberType: PrimitiveType()
+sealed class DataStructureType(open var internalType: Type, open val methods: HashMap<String, DataStructureMethod>): Type() {
+    abstract fun containsMethod(method: String): Boolean
+    abstract fun getMethodByName(method: String): DataStructureMethod
+}
+
+open class DataStructureMethod(open val returnType: Type, open var argumentTypes: List<Type>)
+data class ErrorMethod(override val returnType: Type = NoType, override var argumentTypes: List<Type> = listOf()) : DataStructureMethod(returnType, argumentTypes)
+
+data class StackType(override var internalType: Type = NumberType,
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/frontend/AbstractSyntaxTree.kt:52]: Move all Stack stuff to separate file. This file is v cluttered. Try to make it nicer

File: src/main/kotlin/com/manimdsl/frontend/AbstractSyntaxTreeVisitor.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.manimdsl.frontend
+
+class AbstractSyntaxTreeVisitor {
+    
+    fun visitProgramNode() {
+
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/frontend/AbstractSyntaxTreeVisitor.kt:6]: This has been implemented in #14

File: src/main/kotlin/com/manimdsl/errorhandling/semanticerror/SemanticErrorHandling.kt (hunk 4 above)
Comment (adityagoel4512) [reply src/main/kotlin/com/manimdsl/errorhandling/semanticerror/SemanticErrorHandling.kt:42]: As in the line number isn't in the syntax error listener?

File: src/test/kotlin/com/manimdsl/ValidASTTests.kt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Test
+    fun multiLineProgram() {
+        val multiLineProgram = "let x: number = 1.5;\n" +
+                "# code comment\n" +
+                "let y: Stack = new Stack;\n"
+        val statements = listOf(DeclarationNode(1, "x", NumberNode(1, 1.5)),
+                DeclarationNode(3, "y", ConstructorNode(3, StackType(NumberType), listOf())))
+        val reference = ProgramNode(statements)
+
```
Comment (arj119) [inline src/test/kotlin/com/manimdsl/ValidASTTests.kt:35]: remove this new line

File: src/main/kotlin/com/manimdsl/frontend/Type.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    abstract fun containsMethod(method: String): Boolean
+    abstract fun getMethodByName(method: String): DataStructureMethod
+}
+
+// Pre-defined methods for each data structure
+open class DataStructureMethod(open val returnType: Type, open var argumentTypes: List<Type>)
+data class ErrorMethod(override val returnType: Type = NoType, override var argumentTypes: List<Type> = listOf()) : DataStructureMethod(returnType, argumentTypes)
+
+data class StackType(override var internalType: Type = NumberType,
+                     override val methods: HashMap<String, DataStructureMethod> = hashMapOf("push" to PushMethod(argumentTypes=listOf(NumberType)), "pop" to PopMethod(internalType))): DataStructureType(internalType, methods) {
```
Comment (arj119) [inline src/main/kotlin/com/manimdsl/frontend/Type.kt:19]: This file could get really long maybe consider creating a type package and moving this into a file called StackType.kt

**Comment by arj119** [comment]
Can this be rebased/merged with master and have conflicts resolved please

**Comment by rbanerjee20** [comment]
> Can this be rebased/merged with master and have conflicts resolved please

Still writing tests. Will sort out the conflicts once tests are done

File: src/main/kotlin/com/manimdsl/frontend/Type.kt (hunk 9 above)
Comment (rbanerjee20) [reply src/main/kotlin/com/manimdsl/frontend/Type.kt:23]: `Types` themselves are `ASTNodes` so I could move all of these files into an `ast` package instead?

Comment (rbanerjee20) [reply src/main/kotlin/com/manimdsl/frontend/Type.kt:23]: Also data classes can't inherit from sealed classes if they're placed in different files. We can change the sealed class to an open one if it gets too long but I'd prefer to keep them in the same file for now

**Comment by rbanerjee20** [comment]
The tests have now been written for semantic analysis and I've sorted out the merge conflicts with master. Please can this be reviewed by some/all of @mayanksharma3 @arj119 @iz0825 ?

**Comment by adityagoel4512** [comment]
> The tests have now been written for semantic analysis and I've sorted out the merge conflicts with master. Please can this be reviewed by some/all of @mayanksharma3 @arj119 @iz0825 ?

Have you run the example.manimdsl file?

**Comment by adityagoel4512** [comment]
28/34 tests fail for me @rbanerjee20

File: src/main/kotlin/com/manimdsl/frontend/Type.kt (hunk 9 above)
Comment (mayanksharma3) [reply src/main/kotlin/com/manimdsl/frontend/Type.kt:22]: Yeah the sealed class point is a good one, we had this in WACC too, probably best to leave it in here

**Comment by mayanksharma3** [comment]
Looks good to me! Testing looks quite comprehensive so I'm happy to accept (unless anyone else has any breaking changes they want to make)

**Review by iz0825 — APPROVED** [review]

**Pull request merged**